File: backend/internal/models/task.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

type TaskStatus string

const (
	StatusToDo       TaskStatus = "to_do"
	StatusInProgress TaskStatus = "in_progress"
	StatusCompleted  TaskStatus = "completed"
	StatusArchived   TaskStatus = "archived"
)

type TaskPriority string

const (
	PriorityLow    TaskPriority = "low"
	PriorityMedium TaskPriority = "medium"
	PriorityHigh   TaskPriority = "high"
)

// RecurrenceFrequency defines how often a recurring task repeats.
type RecurrenceFrequency string

const (
	RecurrenceNone     RecurrenceFrequency = ""
	RecurrenceDaily    RecurrenceFrequency = "daily"
	RecurrenceWeekdays RecurrenceFrequency = "weekdays"
	RecurrenceWeekly   RecurrenceFrequency = "weekly"
	RecurrenceMonthly  RecurrenceFrequency = "monthly"
)

// RecurrenceRule holds the repeat schedule for a task.
type RecurrenceRule struct {
	Frequency RecurrenceFrequency `bson:"frequency" json:"frequency"`
	// Until is an optional end-date for the recurrence; nil means repeat forever.
	Until *time.Time `bson:"until,omitempty" json:"until,omitempty"`
}

type Task struct {
	ID bson.ObjectID `bson:"_id,omitempty" json:"id"`

	Title       string       `bson:"title" json:"title"`
	Description string       `bson:"description" json:"description"`
	Status      TaskStatus   `bson:"status" json:"status"`
	Priority    TaskPriority `bson:"priority" json:"priority"`

	DueDate      *time.Time    `bson:"due_date,omitempty" json:"due_date"`
	CreatedAt    time.Time     `bson:"created_at" json:"created_at"`
	UserID       bson.ObjectID `bson:"user_id"`
	ReminderSent bool          `bson:"reminder_sent" json:"reminder_sent"`

	// Recurrence holds the repeat rule, if this is a recurring task.
	Recurrence *RecurrenceRule `bson:"recurrence,omitempty" json:"recurrence,omitempty"`
	// RecurrenceParent is the ID of the original (parent) recurring task, set on child occurrences.
	RecurrenceParent *bson.ObjectID `bson:"recurrence_parent,omitempty" json:"recurrence_parent,omitempty"`
}

type TaskDTO struct {
	Title       string          `bson:"title" json:"title"`
	Description string          `bson:"description" json:"description"`
	Priority    TaskPriority    `bson:"priority" json:"priority"`
	DueDate     *time.Time      `bson:"due_date,omitempty" json:"due_date,omitempty"`
	Recurrence  *RecurrenceRule `bson:"recurrence,omitempty" json:"recurrence,omitempty"`
}

func (dto TaskDTO) ToTask(userID bson.ObjectID) Task {
	return Task{
		Title:       dto.Title,
		Description: dto.Description,
		Status:      StatusToDo,
		Priority:    dto.Priority,
		DueDate:     dto.DueDate,
		CreatedAt:   time.Now(),
		UserID:      userID,
		Recurrence:  dto.Recurrence,
	}
}

type TaskFilter struct {
	Status   TaskStatus   `json:"status"`
	Priority TaskPriority `json:"priority"`
	Search   string       `json:"search"`
	Page     int          `json:"page"`
	PageSize int          `json:"page_size"`
	SortBy   string       `json:"sort_by"`
	SortDir  int          `json:"sort_dir"`
}

type PaginatedResponse struct {
	Tasks      []Task `json:"tasks"`
	Total      int64  `json:"total"`
	Page       int    `json:"page"`
	PageSize   int    `json:"page_size"`
	TotalPages int    `json:"total_pages"`
}

File: frontend/src/types/task.ts
export type TaskStatus = "to_do" | "in_progress" | "completed" | "archived";
export type TaskPriority = "low" | "medium" | "high";
export type RecurrenceFrequency = "daily" | "weekdays" | "weekly" | "monthly" | "";

export type RecurrenceRule = {
  frequency: RecurrenceFrequency;
  until?: string | null;
};

export type Task = {
  id: string;
  title: string;
  description: string;
  status: TaskStatus;
  priority: TaskPriority;
  due_date?: string | null;
  created_at: string;
  recurrence?: RecurrenceRule | null;
  recurrence_parent?: string | null;
};

export type PaginatedTasksResponse = {
  tasks: Task[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
};

export type TaskFormData = {
  title: string;
  description: string;
  priority: TaskPriority;
  status: TaskStatus;
  due_date: string;
  recurrence: RecurrenceFrequency;
};

export type CreateTaskPayload = {
  title: string;
  description: string;
  priority: TaskPriority;
  due_date?: string;
  recurrence?: { frequency: RecurrenceFrequency } | null;
};

export type UpdateTaskPayload = {
  title: string;
  description: string;
  priority: TaskPriority;
  status: TaskStatus;
  due_date: string | null;
  recurrence?: { frequency: RecurrenceFrequency } | null;
};

export const defaultTaskForm: TaskFormData = {
  title: "",
  description: "",
  priority: "medium",
  status: "to_do",
  due_date: "",
  recurrence: "",
};

export const statusLabel: Record<TaskStatus, string> = {
  to_do: "Pending",
  in_progress: "In Progress",
  completed: "Completed",
  archived: "Archived",
};

export const recurrenceLabel: Record<RecurrenceFrequency, string> = {
  "":         "Does not repeat",
  daily:      "Daily",
  weekdays:   "Weekdays (Mon–Fri)",
  weekly:     "Weekly",
  monthly:    "Monthly",
};